### ISSUE TITLE
Questions on Pre-training and pose matrix h

### ISSUE THREAD

**Issue by arekavandi** [vislearn/dsacstar#20]
Hi Eric
Thanks for your wonderful work on pose estimation.
(1). I was checking the performance improvement on the test dataset in the Cambridge_GreatCourt scene starting with your provided initial network and then the trained network in an end-to-end fashion. Surprisingly, the error rates are pretty the same before and after end-to-end learning. In particular, I'm running test.py by
python test.py Cambridge_GreatCourt models\rgb\cambridge\Cambridge_GreatCourt.net --mode 1 
and the results are
![image](https://user-images.githubusercontent.com/101369948/174682693-3a9a8c24-c696-4ff5-87fc-396eaad921b3.png)
However when I train the network for a few epochs in the second phase, by
python train_e2e.py Cambridge_GreatCourt models\rgb\cambridge\Cambridge_GreatCourt.net output\output.net --mode 1
the results are 
![image](https://user-images.githubusercontent.com/101369948/174682828-cefbf841-941f-42d5-9adb-a33f4ff95a25.png)
which means there is no benefit in the end-to-end learning stage.  Is that what you observed before? Or I'm doing something wrong.

(2). I'm a bit confused about the notations in the paper and what has been implemented in the code. In the paper, it was defined that h (4*4 matrix) is the pose matrix from the camera to the 3d scene coordinate. However, the ground truth pose is the matrix transforming from the environment to the camera coordinate. So direct differencing is not correct, right? I do not know within the c++ code whether you are computing the inverse or not, but I can see at the end, you are saving the inverse pose in the output text file. Can you please explain this?

Best regards
Aref

**Comment by ebrach** [comment]
Hi Aref!

Thanks for your message, happy to clarify!

1) The pre-trained networks underwent all training stages, including end-to-end training. So what you observe is that further end-to-end training has no effect if the network was already trained end-to-end.

2) Pose conventions are a constant source of joy :) 
- The paper and the code use the convention that the pose converts camera coordinates to world (=scene) coordinates. If the code is not consistent everywhere wrt this I'm sorry. It should be. But eg [here in the dataset class](https://github.com/vislearn/dsacstar/blob/25accc6ae7b3a446ca86ce2487fbfd3daa0a640a/dataset.py#L249) the pose is setup like that.
- If you want to calculate a reprojection error, you have to invert the pose. That is also the convention used by OpenCV. Therefore, some parts of the code, especially the bits in C++ operate on inverted poses.
- The pose loss in C++ still makes sure that we calculate the pose error with non-inverted poses. The derivatives of the pose loss take the inversion into account.
- It is true that we write out the inverted pose for our estimates. This is to conform with other pipelines - I think COLMAP expects inverted poses, but not completely sure anymore.

Hope that helps,
Eric

**Comment by arekavandi** [comment]
Thanks for your clarification.
Assuming h is a 4 * 4 matrix which maps a point in a given image to the world, 
![image](https://user-images.githubusercontent.com/101369948/178648417-cfc44593-96b4-427e-aee9-df97f8232543.png)
in eq(9) of the paper is doing the reverse, right?
My only problem is the dimension mistmatching in this equation. Since K is 3 * 3 and h inverse is 4 * 4. Can you elaborate on this?
However if we assume h is a pose matrix from the world to the camera, normally in 3D to 2D, h matrix is 3 * 4 and p=KHy has no problem.

**Comment by ebrach** [comment]
Hi Aref,

yeah, you are right. This is an error (or sloppyness) on my part. For notational convenience (or lazyness) I do not differentiate homogeneous and non-homogenous coordinates throughout the paper. In this case, h would in fact be 3x4 where in other Equations of the paper in might need to be 4x4. Sorry for the confusion.

Best,
Eric
